### ISSUE TITLE
Unable to reproduce experimental results in the CoVoST-v2 paper

### ISSUE THREAD

**Issue by nguyenbh** [pytorch/fairseq#3364]
## ❓ Unable to reproduce experimental results in the CoVoST-v2 paper

I have tried to replicate Facebook's  results in the [CoVoST-2 paper ](https://arxiv.org/pdf/2007.10310.pdf)
I followed all commands in [the CoVoST guide ](https://github.com/pytorch/fairseq/blob/master/examples/speech_to_text/docs/covost_example.md)
however, I got results always lower than the paper's results 

More specifically, all experiments used the FB English ASR pre-trained Encoder. 

* French-English: 

  + FB reported (table 2): Fr->En ST training: **26.3**
  + Reproduce: Fr->En ST training: **25.18**

* Spanish-English

  + FB reported (table 2):  Es->En ST training: **23.0**
  + Reproduce: Es->En ST training: **20.43**

It is not very clear to me why I cannot reproduce your results. 

Any suggestions?

**Comment by jmp84** [comment]
@nguyenbh thanks for the report. Could you paste the training logs? Could you also send the git commit you're at?

**Comment by nguyenbh** [comment]
@jmp84 Thank you for your response!

The training and test logs for Spanish-English experiments [train.test.log.zip](https://github.com/pytorch/fairseq/files/6159284/train.test.log.zip)
I am at commit 16c1a200f87a2adb6395e353345c19bbe990d1dd.

**Comment by jmp84** [comment]
@nguyenbh thanks for the logs. Could you also check the BLEU score you are getting when downloading the pre-trained checkpoints? (to make sure the data setup is correct) In addition, can you also try evaluating checkpoint_best.pt that you obtain with your training? Finally, i see in your training log `load_pretrained_encoder_from='/home/nguyen.bach/alitranx/e2e_st/covost/en/asr_en_model/covost2_en_asr_transformer_s.pt'` Could you give details on how this pretrained encoder checkpoint was obtained? Was it the checkpoint_best.pt from the ASR step or was it the averaged checkpoint?

**Comment by nguyenbh** [comment]
@jmp84 
Download [pretrained checkpoint](https://dl.fbaipublicfiles.com/fairseq/s2t/covost2_es_en_st_transformer_s.pt), decode the test set, and I got  **21.77** 
Log file [test.fb_model.log.zip](https://github.com/pytorch/fairseq/files/6159608/test.fb_model.log.zip)

The English ASR is the Facebook pretrained model downloaded [here](https://dl.fbaipublicfiles.com/fairseq/s2t/covost2_en_asr_transformer_s.pt)

**Comment by jmp84** [comment]
@nguyenbh thanks for the info. I recommend following the instructions end to end, including running the ASR pre-training step yourself. I think @kahne was looking into re-uploading the checkpoints after a bug fix in hyperparameters.

**Comment by pyyush** [comment]
@nguyenbh can you try training w/ update_freq = 1, label_smoothing  = 0.1, max_update = 70000, validate_interval = 5, and adam_betas = (0.9, 0.98) ?

**Comment by kahne** [comment]
@nguyenbh Thanks for reporting the issue! We've updated the recipes and pre-trained checkpoints. Please pull the latest master branch and re-download the checkpoint files.

**Comment by nguyenbh** [comment]
@kahne @jmp84 Thanks for following up with this issue.  The recent update of transformer.py seems to be problematic from lines 888-893.

I modify them to

```
if (encoder_out is not None:
    if len(encoder_out["encoder_out"]) > 0:
        enc = encoder_out["encoder_out"][0]
    assert (
        enc.size()[1] == bs
    ), f"Expected enc.shape == (t, {bs}, c) got {enc.shape}"
    if len(encoder_out["encoder_padding_mask"]) > 0:
        padding_mask = encoder_out["encoder_padding_mask"][0]
```

After this modification, I confirm that with the updated ST pretrained checkpoint I can reproduce the Es-En score of 23.1.

**Comment by jmp84** [comment]
@nguyenbh thanks for confirming! I'm going to close this issue and I'm following up on the missing length check.

**Comment by gwenzek** [comment]
Sorry for the breakage, this should have been fixed in ee0d5a0f65a25e5f5372776402aac5cb9c4adbf1
